crates/flux-middle/src/rty/expr.rs

```diff
@@ -611,6 +611,10 @@ impl Expr {
                 let val = const_eval::scalar_int_to_rty_constant2(tcx, *scalar, ty).unwrap();
                 Expr::constant(val)
             }
+            ConstKind::Unevaluated(ue) => {
+                let val = const_eval::unevaluated_to_constant(tcx, ue).unwrap();
+                Expr::constant(val)
+            }
```

Comment:
@ranjitjhala This works as a quick hack, but the proper way to do it (following what rustc does) is to evaluate constants when normalizing signature/types, i.e., the implementation of `normalize_projections` should also normalize/evaluate constants (see https://github.com/rust-lang/rust/blob/2c93fabd98d2c183bcb3afed1f7d51b2517ac5ed/compiler/rustc_trait_selection/src/traits/normalize.rs#L415)

Reply:
Sounds good - how does that work for structs? Do we normalize their types somewhere? Maybe in conv?

Reply:
if a function signature contains a struct with const arguments we will normalize it when we normalize the signature.

Reply:
Hmm I think that -- normalizing the struct/field signatures -- is perhaps not happening (or needs to happen somewhere).

I added your suggestion -- to evaluate the const during `normalization` but the following test fails:

```rust
// #[flux::constant] Add this line if you want to refer to `BUFLEN` in *refinement* specifications
const BUFLEN: usize = 100;

struct Blob {
    data: [i32; BUFLEN],
}

pub fn test(buf: &[i32; BUFLEN]) -> i32 {
    let x0 = buf[0];
    let x1 = buf[10];
    let x2 = buf[BUFLEN - 1];
    let xbad = buf[BUFLEN]; //~ ERROR assertion might fail
    x0 + x1 + x2 + xbad
}

pub fn test_blob(blob: Blob) -> i32 {
    let x0 = blob.data[0];
    let x1 = blob.data[10];
    let x2 = blob.data[BUFLEN - 1];
    let xbad = blob.data[BUFLEN]; //~ ERROR assertion might fail
    x0 + x1 + x2 + xbad
}
```

[I assume the normalization should ensure that we have _no_ `UnevaluatedConst` left over, e.g. when actually checking the `rvalue`?]

Reply:
In short, I think there is some spot where `normalization` is not happening, presumably when we lookup the field `.data` ?

Reply:
Hmm maybe a `normalize` in `downcast_struct`?

Reply:
hmm good point. Normalizing in `downcast_struct` would work (I think), but that sounds a bit wasteful because it will always normalize to the same. I wonder what rustc does in this case 🤔

Reply:
(btw `#flux::constant` is no longer necessary after https://github.com/flux-rs/flux/pull/682 (the attribute has no effect now))

Reply:
ah rustc doesn't have `downcast_struct`, they don't even have an analogous to `VariantSig`. I think they normalize everything when lowering into `mir` and they even annotate fields projection with a (presumably) normalized type (https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/type.PlaceElem.html#variant.Field). So yeah, we either

1. Normalize each time in `downcast_struct` (and `downcast_enum`)
2. Make this normalization into a query to cache it
3. Normalization `VariantSig` as part of `conv` 

3 sounds like the easiest and would likely be more complete. I don't know if there are going to be problems down the road for not storing the non-normalized version.

Reply:
If we do it during `conv` we need to do it after we check structural compatibility

Reply:
hmm I take it back. From your message in Slack I realize that to normalize we need the information at the call site so we need to do it in `downcast_struct`, so it is not that wasteful